msteams-platform/tabs/how-to/dashboard-teams-toolkit.md

```diff
@@ -0,0 +1,991 @@
+      </div>
+    );
+  }
+
+  /**
+   * Get data required by the widget, you can get data from a api call or static data stored in a file. Override this method according to your needs.
+   * @returns data for the widget
+   */
+  protected async getData(): Promise<T> {
+    return new Promise<T>(() => {});
+  }
+
+  /**
+   * Override this method to customize the widget header.
+   * @returns JSX component for the widget body
+   */
+  protected headerContent(): JSX.Element | undefined {
+    return undefined;
+  }
+
+  /**
+   * Override this method to customize the widget body.
+   * @returns JSX component for the widget body
+   */
+  protected bodyContent(): JSX.Element | undefined {
+    return undefined;
+  }
+
+  /**
+   * Override this method to customize the widget footer.
+   * @returns react node for the widget footer
+   */
+  protected footerContent(): JSX.Element | undefined {
+    return undefined;
+  }
+}
+```
+
+Following are the recommended methods to override:
+
+| Methods | Function | Recommend to override |
+|---|---|---|
+| **constructor()** | Invokes the initial `this.state` and call the constructor of the super class `React` Component. | NO |
+| **componentDidMount()** | Invokes after a component is mounted and assigns a value to the `data` property of the state by calling the `getData()` method. | NO |
+| **render()** | Invokes each time a update happens and the dashboard default layout is defined in this method. | NO |
+| **getData()** | Invokes the data needed by the widget, and the value returned by this method will be set to `this.state.data`. |
+| **headerContent()** | Invokes what the widget header will look like. You can choose to override this method to customize a widget or not, if not, the widget won't have a header. | YES |
+| **bodyContent()** | Invokes what the widget body will look like. You can choose to override this method to customize a widget or not, if not, the widget won't have a body. | YES |
+| **footerContent()** | Invokes what the widget footer will look like. You can choose to override this method to customize a widget or not, if not, the widget won't have a footer. | YES |
+
+### Microsoft Graph Toolkit as widget content
+
+Microsoft Graph Toolkit is a set of renewable, framework-agnostic web component which helps accessing and working with Microsoft Graph. You can use the Microsoft Graph Toolkit with any web framework or without a framework.
+
+You can follow the steps to use Microsoft Graph Toolkit as your widget content:
+
+1. Add SSO feature to your Teams app
+
+   Microsoft Teams provides single sign-on (SSO) function for an app to obtain signed in Teams user token to access Microsoft Graph.
+
+   For more information, refer [SSO feature to your Teams app](../../toolkit/add-single-sign-on.md).
+
+1. Install required `npm` packages
+
+   Run the following command in your project `tabs` folder to install the required `npm` packages:
+
+      ```typescript
```

Comment:
The language of this code snippet should be `Bash`.

Reply:
Updated.